internal/services/groups/groups_data_source.go

```diff
@@ -62,13 +69,27 @@ func groupsDataSourceRead(ctx context.Context, d *schema.ResourceData, meta inte
 
 	var groups []msgraph.Group
 	var expectedCount int
+	var returnAll = d.Get("return_all").(bool)
 
 	var displayNames []interface{}
 	if v, ok := d.GetOk("display_names"); ok {
 		displayNames = v.([]interface{})
 	}
 
-	if len(displayNames) > 0 {
+	if returnAll {
+		result, _, err := client.List(ctx, odata.Query{})
+		if err != nil {
+			return tf.ErrorDiagF(err, "Could not retrieve groups")
+		}
+		if result == nil {
+			return tf.ErrorDiagF(errors.New("API returned nil result"), "Bad API Response")
+		}
+		if len(*result) == 0 {
+			return tf.ErrorDiagPathF(err, "return_all", "No groups found")
+		}
+
+		groups = append(groups, *result...)
```

Comment:
This notation is a useful shortcut to iterating with a for loop

Reply:
Didn't know this one - Thank you for that - I will read through all these comments :)